File: v1/LLD/Design-Patterns/Observer/code/observer.py
"""
Observer Pattern — a StockTicker (Subject) notifies registered Observers on price change.
Run: python observer.py
"""

from abc import ABC, abstractmethod


class StockObserver(ABC):
    @abstractmethod
    def update(self, symbol, price):
        ...


class StockTicker:
    """The Subject — maintains observers and notifies them on state change."""

    def __init__(self, symbol):
        self.symbol = symbol
        self.price = 0.0
        self._observers = []

    def attach(self, observer: StockObserver):
        self._observers.append(observer)

    def detach(self, observer: StockObserver):
        self._observers.remove(observer)

    def set_price(self, new_price):
        self.price = new_price
        self._notify_observers()

    def _notify_observers(self):
        for observer in self._observers:
            observer.update(self.symbol, self.price)


class MobileDisplay(StockObserver):
    def update(self, symbol, price):
        print(f"[MobileDisplay] {symbol} is now ${price}")


class EmailAlert(StockObserver):
    def __init__(self, threshold):
        self.threshold = threshold

    def update(self, symbol, price):
        if price >= self.threshold:
            print(f"[EmailAlert] {symbol} crossed ${self.threshold} — now ${price}")


if __name__ == "__main__":
    ticker = StockTicker("ACME")

    mobile_display = MobileDisplay()
    email_alert = EmailAlert(100.0)

    ticker.attach(mobile_display)
    ticker.attach(email_alert)

    print("Price update 1:")
    ticker.set_price(95.0)

    print("Price update 2:")
    ticker.set_price(102.5)

    print("Detaching EmailAlert.")
    ticker.detach(email_alert)

    print("Price update 3:")
    ticker.set_price(110.0)

File: v1/LLD/Design-Patterns/Observer/code/Observer.java
// Observer Pattern — a StockTicker (Subject) notifies registered Observers on price change.
// Compile: javac Observer.java
// Run:     java Observer

import java.util.ArrayList;
import java.util.List;

interface StockObserver {
    void update(String symbol, double price);
}

class StockTicker {
    private List<StockObserver> observers = new ArrayList<>();
    private String symbol;
    private double price;

    StockTicker(String symbol) {
        this.symbol = symbol;
    }

    void attach(StockObserver observer) {
        observers.add(observer);
    }

    void detach(StockObserver observer) {
        observers.remove(observer);
    }

    void setPrice(double newPrice) {
        this.price = newPrice;
        notifyObservers();
    }

    private void notifyObservers() {
        for (StockObserver observer : observers) {
            observer.update(symbol, price);
        }
    }
}

class MobileDisplay implements StockObserver {
    public void update(String symbol, double price) {
        System.out.println("[MobileDisplay] " + symbol + " is now $" + price);
    }
}

class EmailAlert implements StockObserver {
    private double threshold;

    EmailAlert(double threshold) {
        this.threshold = threshold;
    }

    public void update(String symbol, double price) {
        if (price >= threshold) {
            System.out.println("[EmailAlert] " + symbol + " crossed $" + threshold + " — now $" + price);
        }
    }
}

public class Observer {
    public static void main(String[] args) {
        StockTicker ticker = new StockTicker("ACME");

        MobileDisplay mobileDisplay = new MobileDisplay();
        EmailAlert emailAlert = new EmailAlert(100.0);

        ticker.attach(mobileDisplay);
        ticker.attach(emailAlert);

        System.out.println("Price update 1:");
        ticker.setPrice(95.0);

        System.out.println("Price update 2:");
        ticker.setPrice(102.5);

        System.out.println("Detaching EmailAlert.");
        ticker.detach(emailAlert);

        System.out.println("Price update 3:");
        ticker.setPrice(110.0);
    }
}
